tempesta_fw/client.c

```diff
@@ -28,122 +28,222 @@
 #include "connection.h"
 #include "log.h"
 #include "procfs.h"
+#include "tdb.h"
+#include "lib/str.h"
+#include "lib/common.h"
 
-#define CLI_HASH_BITS	17
-#define CLI_HASH_SZ	(1 << CLI_HASH_BITS)
+/* Length of comparison of clients entry by User-Agent. */
+#define UA_CMP_LEN	256
 
-typedef struct {
-	struct hlist_head	list;
-	spinlock_t		lock;
-} CliHashBucket;
+static struct {
+	unsigned int	db_size;
+	const char	*db_path;
+	unsigned int	lifetime;
+} client_cfg __read_mostly;
 
-/*
- * TODO probably not the best container for the task and
- * HTrie should be used instead.
+/**
+ * Client tdb entry.
+ *
+ * @cli			- client descriptor
+ * @expires		- expiration time for the client descriptor after all
+ *			  connections are closed
+ * @lock		- lock for atomic change @expires and @users;
+ * @users		- reference counter.
+ * 			  Expiration state will begind, when the counter reaches
+ *			  zero
  */
-CliHashBucket cli_hash[CLI_HASH_SZ] = {
-	[0 ... (CLI_HASH_SZ - 1)] = {
-		HLIST_HEAD_INIT,
-	}
-};
+typedef struct {
+	TfwClient	cli;
+	TfwAddr		cli_addr;
```

Comment:
`cli_addr` can be misleading, since `cli->addr` exists. The field describes `X-Forwarded-For` address, should we name it `xff_addr` instead?